### DISCUSSION TITLE
Can't make work example play audio from SDFAT to BlueToot

### DISCUSSION THREAD

**Discussion by Marco1971Repo** [pschatzmann/arduino-audio-tools#792] [Q&A] [answered]
I'm trying to play sdfta to my bluetooth hedphone without success.
The sample [player-sdfat-audiokit] work as expected.
No way  (for me obviously probably i'm doing something wrong but I don't understand what) to make work player-sdfat-a2dp..
Then I tride to change [player-sdfat-audiokit] removing kit as an output and adding A2DPStream (below the sketch) and the result is a failure initializing SD and i don't understand why

[I] AudioA2DP.h : 141 - a2dp_source is connected...
[I] AudioPlayer.h : 351 - setVolume(0.700000)
[I] VolumeStream.h : 191 - setVolume: 0.700000
[I] VolumeStream.h : 191 - setVolume: 0.700000
[D] AudioPlayer.h : 149 - virtual bool audio_tools::AudioPlayer::begin(int, bool)
[D] AudioStreams.h : 52 - virtual void audio_tools::AudioStream::setAudioInfo(audio_tools::AudioInfo)
[I] AudioTypes.h : 78 - sample_rate: 44100
[I] AudioTypes.h : 79 - channels: 2
[I] AudioTypes.h : 80 - bits_per_sample: 16
[D] AudioEncoded.h : 316 - virtual bool audio_tools::EncodedAudioPrint::begin()
[D] CodecMP3Helix.h : 79 - virtual void audio_tools::MP3DecoderHelix::begin()
[D] MetaDataFilter.h : 30 - void audio_tools::MetaDataFilter<Decoder>::begin() [with Decoder = libhelix::MP3DecoderHelix]
[D] AudioSourceSDFAT.h : 75 - virtual void audio_tools::AudioSourceSDFAT::begin()
[E] AudioSourceSDFAT.h : 78 - sd.begin failed
[I] MetaDataID3.h : 564 - virtual void audio_tools::MetaDataID3::begin()
[I] AudioSourceSDFAT.h : 101 - [D] AudioA2DP.hselectStream SDFAT: 0
 : [E] AudioSourceSDFAT.h305 -  : 111 - Filename is null
[W] AudioPlayer.h : 185 - -> begin: no data found_data: 512 -> 512
> begin: no data found
[D] AudioA2DP.h : 305 - a2dp_stream_source_sound_data: 512 -> 512
`
#include "AudioTools.h"
#include "AudioLibs/AudioKit.h"
#include "AudioLibs/AudioSourceSDFAT.h" // or AudioSourceIdxSDFAT.h
#include "AudioCodecs/CodecMP3Helix.h"
#include "AudioLibs/AudioA2DP.h"
const char *startFilePath="/";
const char* ext="mp3";
SdSpiConfig sdcfg(PIN_AUDIO_KIT_SD_CARD_CS, DEDICATED_SPI, SD_SCK_MHZ(10) , &AUDIOKIT_SD_SPI);
AudioSourceSDFAT source(startFilePath, ext, sdcfg);
//AudioKitStream kit;
MP3DecoderHelix decoder;  // or change to MP3DecoderMAD
//AudioPlayer player(source, kit, decoder);

A2DPStream out;
AudioPlayer player(source, out, decoder);

void next(bool, int, void*) {
   player.next();
}

void previous(bool, int, void*) {
   player.previous();
}

void startStop(bool, int, void*) {
   player.setActive(!player.isActive());
}

void setup() {
  Serial.begin(115200);
  AudioLogger::instance().begin(Serial, AudioLogger::Debug);

  /*// setup output
  auto cfg = kit.defaultConfig(TX_MODE);
  kit.begin(cfg);*/
  auto cfg = out.defaultConfig(TX_MODE);
  cfg.name = "Redmi Buds 3 Lite";  // set the device here. Otherwise the next available device is used for output
  //cfg.auto_reconnect = true;  // if this is use we just quickly connect to the last device ignoring cfg.name
  out.begin(cfg);
 /*// setup additional buttons 
  kit.addAction(PIN_KEY1, startStop);
  kit.addAction(PIN_KEY4, next);
  kit.addAction(PIN_KEY3, previous);
*/

  // setup player
  player.setVolume(0.7);
  player.begin();

  // select file with setPath() or setIndex()
  //player.setPath("/ZZ Top/Unknown Album/Lowrider.mp3");
  //player.setIndex(1); // 2nd file

}

void loop() {
  player.copy();
 // kit.processActions();
}`

**Comment by pschatzmann** [comment]
Why don't you use the tested examples.
For problems with the SD drive did you read the Wiki of the AudioKit project ?

**Reply by Marco1971Repo** [reply]
I did but it (player-sdfat-a2dp) fail:

[W] AudioA2DP.h : 271 - ==> state: Connecting
[W] AudioA2DP.h : 271 - ==> state: Connected
[E] AudioSourceSDFAT.h : 78 - sd.begin failed
[E] AudioSourceSDFAT.h : 111 - Filename is null
[W] AudioPlayer.h : 185 - -> begin: no data found

And this is what i don't understand since (player-sdfat-audiokit) work as expected so there is no problem with SD or at least I think so.

**Comment by Marco1971Repo** [comment, marked as answer]
I've solved the problem but onestly I don't understand to make work the player-sdfat-a2dp I need to add the lasta parameter setting it to false:
AudioSourceSDFAT source(startFilePath, ext, sdcfg, false);

The player-sdfat-audiokit work with the defalut value true....

Maybe you understand why and if it is a bug or not.

Hope is usefull for somebody else :-)

Thanks for your nice work.

**Comment by pschatzmann** [comment]
Hmm, this does not make any sense: The last parameter is just there so that the API is consistent with the AudioSourceIdxSDFAT class. However it is not used!

**Reply by Marco1971Repo** [reply]
I agree with you but....this is it.

**Comment by Marco1971Repo** [comment]
To close this: I was not aligned to yor last commit.
Once aligned arduino-audiokit and arduino-audio-tools to your last commit everything has worked as expected.
